File: services/signature_service.py
import os
from dropbox_sign import ApiClient, Configuration, apis, models
from dotenv import load_dotenv

load_dotenv()

class SignatureService:
    def __init__(self):
        self.api_key = os.getenv("DROPBOX_SIGN_API_KEY")
        if self.api_key:
            self.configuration = Configuration(api_key={'api_key': self.api_key})
        else:
            self.configuration = None

    def send_signature_request(self, pdf_path, email, name, subject, message):
        """
        Sends a signature request via Dropbox Sign.
        If no API key is set, it simulates the request.
        """
        if not self.configuration:
            return {
                "status": "simulated",
                "request_id": "sim_12345",
                "signing_url": f"https://app.hellosign.com/s/simulated_link_for_{name}"
            }

        with ApiClient(self.configuration) as api_client:
            signature_request_api = apis.SignatureRequestApi(api_client)

            signer_1 = models.SubSignatureRequestSigner(
                email_address=email,
                name=name,
                order=0,
            )

            request_data = models.SignatureRequestSendRequest(
                title=subject,
                subject=subject,
                message=message,
                signers=[signer_1],
                files=[open(pdf_path, "rb")],
                test_mode=True,  # Default to test mode
            )

            try:
                response = signature_request_api.signature_request_send(request_data)
                return {
                    "status": "sent",
                    "request_id": response.signature_request.signature_request_id,
                    "is_complete": response.signature_request.is_complete
                }
            except Exception as e:
                print(f"Error sending signature request: {e}")
                return {"status": "error", "message": str(e)}

# Example usage
if __name__ == "__main__":
    service = SignatureService()
    # Mock call
    result = service.send_signature_request(
        "test_vacation_request.pdf", 
        "user@example.com", 
        "Juan Perez", 
        "Solicitud de Vacaciones", 
        "Por favor firma tu solicitud de vacaciones."
    )
    print(result)

File: services/signatureService.js
const DropboxSign = require("@dropbox/sign");
const fs = require("fs");
const path = require("path");

class SignatureService {
    constructor() {
        this.apiKey = process.env.DROPBOX_SIGN_API_KEY;
        if (this.apiKey) {
            this.apiInstance = new DropboxSign.SignatureRequestApi();
            this.apiInstance.username = this.apiKey;
            
            // Forzar test_mode=1 en todas las solicitudes salientes como workaround
            this.apiInstance.addInterceptor((requestOptions) => {
                // Siempre añadir como query param en la URL para evitar payment_required
                requestOptions.params = requestOptions.params || {};
                requestOptions.params.test_mode = 1;
                return Promise.resolve();
            });
        } else {
            console.warn("WARNING: DROPBOX_SIGN_API_KEY not found. Running in simulation mode.");
            this.apiInstance = null;
        }
    }

    async sendSignatureRequest(applicant_name, applicant_email, pdfPath) {
        if (!this.apiInstance) {
            console.log("SIMULATION: Sending signature request to", applicant_email);
            return { request_id: "sim_" + Date.now(), signing_info: "http://localhost:5000/mock-sign" };
        }

        const signer = {
            name: applicant_name,
            emailAddress: applicant_email
        };

        const data = {
            title: "Solicitud de Vacaciones",
            subject: "Por favor firma tu solicitud de vacaciones",
            message: "Hola, tu solicitud ha sido aprobada. Por favor firmala para completar el proceso.",
            signers: [signer],
            files: [
                {
                    value: fs.readFileSync(pdfPath),
                    options: {
                        filename: 'solicitud_vacaciones.pdf',
                        contentType: 'application/pdf'
                    }
                }
            ],
            testMode: 1,
            useTextTags: true
        };

        try {
            const result = await this.apiInstance.signatureRequestSend(data);
            
            // Log for debugging
            console.log("Dropbox Sign Response:", JSON.stringify(result.body, null, 2));

            return {
                request_id: result.body.signatureRequest.signatureRequestId,
                signing_info: result.body.signatureRequest.signingUrl || "Check your email for the signing link"
            };
        } catch (error) {
            console.error("Exception when calling Dropbox Sign API:", error);
            return { request_id: null, signing_info: null };
        }
    }
}

module.exports = SignatureService;
